pkg/distributor/distributor.go

```diff
@@ -1347,41 +1340,66 @@ func (d *Distributor) push(ctx context.Context, pushReq *Request) error {
 		tenantRing = d.ingestersRing.ShuffleShard(userID, d.limits.IngestionTenantShardSize(userID))
 	}
 
-	// we must not re-use buffers now until all DoBatch goroutines have finished,
-	// so set this flag false and pass cleanup() to DoBatch.
+	// we must not re-use buffers now until all writes to backends (e.g. ingesters) have completed, which can happen
+	// even after this function returns. For this reason, it's unsafe to cleanup in the defer and we'll do the cleanup
+	// once all backend requests have completed (see cleanup function passed to sendWriteRequestToBackends()).
 	cleanupInDefer = false
 
-	err = ring.DoBatchWithOptions(ctx, ring.WriteNoExtend, tenantRing, keys,
+	return d.sendWriteRequestToBackends(ctx, userID, req, keys, initialMetadataIndex, tenantRing, func() {
+		// This cleanup function is called once all requests to all backends have been completed.
+		// At this point is safe to finally release the original pushReq resources back to the pool.
```

Comment:
Nit: I think this comment should be on the `sendWriteRequestToBackends` method.

Reply:
Done in https://github.com/grafana/mimir/pull/7877/commits/d6469326d89e2f0ef5d6628825a303474fd8c49e.